src/java.base/share/classes/javax/crypto/spec/HKDFParameterSpec.java

```diff
@@ -0,0 +1,503 @@
+         * creates a new list if there is none yet.
+         * <p>
+         * This supports the use-case where a label can be applied to the salt
+         * but the actual value of the salt is not yet available. An
+         * implementation should concatenate the salt into a single value once
+         * all components are available.
+         *
+         * @param salt
+         *     the salt value
+         *
+         * @return this builder
+         *
+         * @throws NullPointerException
+         *     if the {@code salt} is null
+         */
+        public Builder addSalt(byte[] salt) {
+            if (salt == null) {
+                throw new NullPointerException(
+                    "salt must not be null or empty");
+            }
+            if (salt.length != 0) {
+                return addSalt(new SecretKeySpec(salt, "Generic"));
+            } else {
+                return this;
+            }
+        }
+    }
+
+    /**
+     * Returns a builder for building {@code Extract} and
+     * {@code ExtractThenExpand} objects.
+     * <p>
+     * Note: one or more of the methods {@code addIKM} or {@code addSalt} should
+     * be called next, before calling build methods, such as
+     * {@code Builder.extractOnly()}
+     *
+     * @return a {@code Builder} to mutate
+     */
+    static Builder ofExtract() {
+        return new Builder().createBuilder();
+    }
+
+    /**
+     * Defines the input parameters of an {@code Expand} object
+     *
+     * @param prk
+     *     the pseudorandom key; must not be {@code null} in the Expand case
+     * @param info
+     *     the optional context and application specific information (may be
+     *     {@code null}); the byte[] is copied to prevent subsequent
+     *     modification
+     * @param length
+     *     the length of the output key material (must be &gt; 0 and &lt; 255 *
+     *     HMAC length)
+     *
+     * @return a new {@code Expand} object
+     *
+     * @throws NullPointerException
+     *     if {@code prk} is {@code null}
+     * @throws IllegalArgumentException
+     *     if {@code length} is not > 0
+     */
+    static Expand expandOnly(SecretKey prk, byte[] info, int length) {
+        if (prk == null) {
+            throw new NullPointerException("prk must not be null");
+        }
+        return new Expand(prk, info, length);
+    }
+
+    /**
+     * Defines the input parameters of an Extract operation as defined in <a
+     * href="http://tools.ietf.org/html/rfc5869">RFC 5869</a>.
+     */
+    @PreviewFeature(feature = PreviewFeature.Feature.KEY_DERIVATION)
+    final class Extract implements HKDFParameterSpec {
+
+        // HKDF-Extract(salt, IKM) -> PRK
+        private final List<SecretKey> ikms;
+        private final List<SecretKey> salts;
+
+        private Extract() {
+            this(new ArrayList<>(), new ArrayList<>());
+        }
+
+        private Extract(List<SecretKey> ikms, List<SecretKey> salts) {
```

Comment:
Should it throw an exception if either or both of the lists are empty?

Reply:
If that's the case, then I think the `ikms` and `salts` method should specify that an empty List may be returned and under what conditions, as it is an edge case. For example, "Returns an empty list if there are no salt values".

Reply:
Resolved.